tempesta_fw/http_parser.c

```diff
@@ -3199,11 +3199,27 @@ tfw_http_parse_terminate(TfwHttpMsg *hm)
 	BUG_ON(!hm);
 	BUG_ON(!(TFW_CONN_TYPE(hm->conn) & Conn_Srv));
 
-	if (hm->parser.state == Resp_BodyUnlimRead) {
+	/*
+	 * Set Content-Length header to warn client about end of message.
+	 * Other option is to close connection to client. All the situations
+	 * when Content-Length header must not present in responce were
+	 * checked earlier. Refer to RFC 7230 3.3.3
+	 */
+	if (hm->parser.state == Resp_BodyUnlimRead
+	    || hm->parser.state == Resp_BodyUnlimStart)
+	{
+		char c_len[20] = {0};
+		int r;
+
 		BUG_ON(hm->body.flags & TFW_STR_COMPLETE);
 		hm->body.flags |= TFW_STR_COMPLETE;
 		hm->content_length = hm->body.len;
-		return true;
+		sprintf(c_len, "%lu", hm->content_length);
```

Comment:
We do not use format strings in Tempesta FW since they're slow. In fact you know the argument type and don't need to run generic format string logic. Secondly do not use length-less functions, i,e, `snprintf()` should be preferred to `sprintf()`. Please add following code (with appropriate kernel API) to str.[ch] with appropriate tests and use the function here.

        #include <stdio.h>
        
        /*
         * Quickly get number of digits - 1, e.g. returns '0' for '7' and '2' for '333'.
         * It's assumed that small numbers are likely.
         */
        static inline unsigned int
        dig_num(unsigned int a)
        {
        	if (a < 10)
        		return 0;
        	if (a < 100)
        		return 1;
        	if (a < 1000) 
        		return 2;
        	if (a < 10000)
        		return 3;
        	if (a < 100000)
        		return 4;
        	if (a < 1000000)
        		return 5;
        	if (a < 10000000)
        		return 6;
        	if (a < 100000000)
        		return 7;
        	if (a < 1000000000)
        		return 8;
        	return 9;
        }
        
        #define unlikely(c)	__builtin_expect((c), 0)
        
        static size_t
        uitoa(unsigned int ai, char *buf, size_t len)
        {
        	size_t n = dig_num(ai);
        	char *p = buf + n;
        
        	if (unlikely(n > len))
        		return 0;
        
        	do {
        		/* Compiled to only one MUL instruction with -O2. */
        		*p-- = "0123456789"[ai % 10];
                	ai /= 10;
        	} while (ai);
        
        	return n + 1;
        }
        
        #define T(a)								\
        do {									\
        	char buf[10];							\
        	size_t n = uitoa(a, buf, 10);					\
        	printf("%u -> %.*s (len=%u)\n", a, n, buf, n);			\
        } while (0)
        
        int
        main(int argc, char *argv[])
        {
        
        	T(0);
        	T(10);
        	T(0xff);
        	T(0xaff);
        	T(0x45ff);
        	T(0xaabbccff);
        	T(0xffffffff);
        
        	return 0;
        }